### ISSUE TITLE
Snapshot creation error: found chunk outside of compacted

### ISSUE THREAD

**Issue by zuzzas** [prometheus/prometheus#5105]
## Bug Report
**What did you do?**
Updated Prometheus v2.5.0 -> v2.6.0.
**What did you expect to see?**
A working snapshotting feature.
**What did you see instead? Under which circumstances?**
Snapshots stopped working, recovered backups from v2.5.0 aren't working as well.

```
{"status":"error","errorType":"internal","error":"create snapshot: snapshot head block: write compaction: found chunk with minTime: 1547726468168 maxTime: 1547726528168 outside of compacted minTime: 1547719200000 maxTime: 1547726527867"}
```
**Environment**

* System information:

	Linux 4.15.0-43-generic x86_64

* Prometheus version:

	prometheus, version 2.6.0 (branch: HEAD, revision: dbd1d58c894775c0788470944b818cc724f550fb)
  build user:       root@bf5760470f13
  build date:       20181217-15:14:46
  go version:       go1.11.3

* Logs:
```
level=info ts=2019-01-17T12:00:00.045764719Z caller=db.go:787 component=tsdb msg="snapshotting block" block=01D0ZBFH0DZABWG7VYNNVTX9V7
level=info ts=2019-01-17T12:00:00.046032635Z caller=db.go:787 component=tsdb msg="snapshotting block" block=01D154WDRC1RJ4Z8BJPYYHCP4C
level=info ts=2019-01-17T12:00:00.046270873Z caller=db.go:787 component=tsdb msg="snapshotting block" block=01D1AYA05AFGWN8AP8QCPBXP7Q
level=info ts=2019-01-17T12:00:00.046494941Z caller=db.go:787 component=tsdb msg="snapshotting block" block=01D1CW28GMGNRRQDGN3DEWF0H9
level=info ts=2019-01-17T12:00:00.046644839Z caller=db.go:787 component=tsdb msg="snapshotting block" block=01D1DGMT7Q571MBSV66MG2M55C
level=info ts=2019-01-17T12:00:00.046755629Z caller=db.go:787 component=tsdb msg="snapshotting block" block=01D1DGKQWN3Q4K870EQTGCJ0CD
level=info ts=2019-01-17T12:00:00.046845693Z caller=db.go:787 component=tsdb msg="snapshotting block" block=01D1DQFF5Q5EXDTT5GRKB470SD
level=info ts=2019-01-17T12:00:27.62712416Z caller=db.go:787 component=tsdb msg="snapshotting block" block=01D0ZBFH0DZABWG7VYNNVTX9V7
level=info ts=2019-01-17T12:00:27.627376645Z caller=db.go:787 component=tsdb msg="snapshotting block" block=01D154WDRC1RJ4Z8BJPYYHCP4C
level=info ts=2019-01-17T12:00:27.627594562Z caller=db.go:787 component=tsdb msg="snapshotting block" block=01D1AYA05AFGWN8AP8QCPBXP7Q
level=info ts=2019-01-17T12:00:27.627812374Z caller=db.go:787 component=tsdb msg="snapshotting block" block=01D1CW28GMGNRRQDGN3DEWF0H9
level=info ts=2019-01-17T12:00:27.627949074Z caller=db.go:787 component=tsdb msg="snapshotting block" block=01D1DGMT7Q571MBSV66MG2M55C
level=info ts=2019-01-17T12:00:27.628038373Z caller=db.go:787 component=tsdb msg="snapshotting block" block=01D1DGKQWN3Q4K870EQTGCJ0CD
level=info ts=2019-01-17T12:00:27.628112874Z caller=db.go:787 component=tsdb msg="snapshotting block" block=01D1DQFF5Q5EXDTT5GRKB470SD
```

**Comment by simonpasquier** [comment]
What flags do you pass to Prometheus? Can you try using the [`tsdb`](https://github.com/prometheus/tsdb/tree/master/cmd/tsdb) CLI to check the data directory?

**Comment by zuzzas** [comment]
Will try tsdb tool.

```
/bin/prometheus --storage.tsdb.retention=1008h --storage.tsdb.path=/var/prometheus/data --config.file=/etc/prometheus/config/prometheus.yaml --web.route-prefix=/prometheus/ --web.external-url=https://test.flant.com/prometheus/ --web.console.libraries=/etc/prometheus/console_libraries --web.console.templates=/etc/prometheus/consoles --web.enable-lifecycle --web.enable-admin-api --rules.alert.resend-delay=30s```

**Comment by gouthamve** [comment]
Hi @zuzzas, can you elaborate on "Snapshots stopped working"? I am able to snapshot and recover just fine on `2.6.1` which afaics has the same TSDB code as `2.6`.

I'm looking into the restore from 2.5

**Comment by gouthamve** [comment]
Ah, I see the error now, sorry for missing it earlier.

**Comment by zuzzas** [comment]
@gouthamve 
I've `rm -rf`'ed the entire Prometheus data directory. It still fails with the same error. I've tried turning on the debug log level, but there is nothing new. Could there be some strange metric that causes TSDB to misbehave?

**Comment by zuzzas** [comment]
I've resolved it by passing `?skip_head=true` URL option to snapshot command. I lose a couple of hours in backup, but it works. Still, far from ideal.

**Comment by bwplotka** [comment]
Looks like https://github.com/prometheus/tsdb/pull/514 should fix this.

**Comment by bwplotka** [comment]
Hi guys, we are quite busy on other stuff and the fix I had has was not fixing all theoretical problems we can have.

We need some feedback to prioritze this well. Are you still affected? How painful it is?

**Comment by zuzzas** [comment]
I'll update our affected deployment to the latest Prometheus and remove `?skip_head=true` from the request, so that I can test whether we are still affected.

**Comment by bentsi** [comment]
We hit this issue on official Prometheus v2.7.2 Docker
```python
{u'status': u'error', u'errorType': u'internal', u'error': u'create snapshot: snapshot head block: write compaction: found chunk with minTime: 1560511080568 maxTime: 1560511140568 outside of compacted'}
```

**Comment by bwplotka** [comment]
ACK

**Comment by rjshrjndrn** [comment]
Guys, any update on this?

Because now out production backup is quite flaky.

**Comment by Wsyzxxn** [comment]
I Hava same error 
level=error ts=2019-06-24T01:12:12.016262506Z caller=db.go:338 component=tsdb msg="compaction failed" err="persist head block: write compaction: found chunk with minTime: 1560412868427 maxTime: 1560416530844 outside of compacted minTime: 1560414600000 maxTime: 1560420000000"

drwxr-xr-x 3 root root 131072 Jun 24 09:51 wal
drwxr-xr-x 3 root root   4096 Jun 24 09:49 01DE3JRK97ZHPAD2HRJ0A70KPJ.tmp
drwxr-xr-x 3 root root   4096 Jun 13 16:45 01DD806404SSBE67BSAHX3KADE
drwxr-xr-x 3 root root   4096 Jun 13 16:15 01DD7YF646KJGX540N1JFVSJFK
drwxr-xr-x 3 root root   4096 Jun 13 15:45 01DD7WR8F344GC7KMNE8KJYWJY
drwxr-xr-x 3 root root   4096 Jun 13 15:15 01DD7V1AJ9Z7RPKKP9YT1DTPPG
drwxr-xr-x 3 root root   4096 Jun 13 14:45 01DD7SACTKNSMSJHV4DJD7AEHJ
drwxr-xr-x 3 root root   4096 Jun 13 14:15 01DD7QKEWWYJ9CC902XB0KE2XY
drwxr-xr-x 3 root root   4096 Jun 13 13:45 01DD7NWH4Y1Z70XFB6WTQYFSEH
drwxr-xr-x 3 root root   4096 Jun 13 13:15 01DD7M5KCB6XCFP9408PN4MKVF
drwxr-xr-x 3 root root   4096 Jun 13 12:45 01DD7JENPR9DJF5BT5FMR5N2WY

**Comment by bwplotka** [comment]
Current approach is to dynamically rewrite problematic "shared" (with appender) chunks. Exploring this here: https://github.com/prometheus/tsdb/pull/639

**Comment by zhanglijingisme** [comment]
Hi @bwplotka @brian-brazil
I'm using Prometheus v2.9.1 and hit this today, when I'm trying to create snapshot for data (3.8G size).
{"status":"error","errorType":"internal","error":"create snapshot: snapshot head block: write compaction: found chunk with minTime: 1564020019830 maxTime: 1564023139830 outside of compacted minTime: 1564012800000 maxTime: 1564023138442"}

Seems this still exists in v2.9.1 (released in Apr.17th) - Is there any fix on latest version 2.11.1 (released in Jul. 10th)?

**Comment by krasi-georgiev** [comment]
Yes the fix is in 2.11.1.

Locking this issue, but feel free to open a new if you still experience the same issues after the upgrade.
